### PR TITLE
Junit reporter

### PR DESCRIPTION
**Pull request by lbartnik**
Hi @jspitzen, @hadley 

I've rewritten the JUnit reporter to match the new API. I understand xml2 and its own tests are missing, I'll work on that in the coming days. In the meantime, could you please let me know if there is anything missing/flawed as far as JUnit/testthat designs go?

I want to integrate testthat with JUnit/XML output but I'm not a JUnit expert.

Lukasz

### CODE REVIEW

**Comment by lbartnik** [comment]
Hi, @hadley  given your [comment](https://github.com/hadley/testthat/pull/260#discussion_r40376020) I have taken a look at `xml2` and it seems that this package does not have a convenient API to build a XML document the way you can do it with `XML`. Any hints on how to replace `XML` with `xml2` in this case?

**Comment by hadley** [comment]
It will have it very soon - see https://github.com/hadley/xml2/blob/master/vignettes/modification.Rmd for how it works in the dev version.

**Comment by lbartnik** [comment]
I actually have. It seems one needs to create an empty document and then modify it, is that right? So make we could make it another pr when xml2 is ready and make this the first version?

**Comment by hadley** [comment]
Since xml2 is now out, would you mind taking another look at this PR? It's much appreciated!

**Comment by lbartnik** [comment]
I'll take a look.

File: tests/testthat/test-reporter-junit.R [hunk 1]
```diff
@@ -0,0 +1,6 @@
+context("JUnitReporter")
+
+test_that("Junit reporter produces XML", {
+
```
Comment (hadley) [inline tests/testthat/test-reporter-junit.R]: Needs some sort of regression test here

File: R/reporter-junit.R [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        ## do XML entity substitutions
+        if (!is.null(attrs)) {
+          attrs <- vapply(attrs, function(x) toString(xmlTextNode(x)), character(1))
+        }
+        xmlNode(name, ..., attrs = attrs)
+      }
+      classnameOK <- function(text) {
+        gsub("[ \\.]", "_", text)
+      }
+      suites <- lapply(names(self$results), function(context) {
```
Comment (hadley) [inline R/reporter-junit.R]: I think this would be easier to follow if you pulled out separate functions for the expectation, test, and context levels

Comment (lbartnik) [inline R/reporter-junit.R]: @hadley I hope it's not a blocker? I've just pulled @jspitzen's PR and translated to new testhat API. I'd rather not put more work into that, given your request to switch to `xml2`

File: tests/testthat/test-reporter-junit.R (hunk 1 above)
Comment (lbartnik) [inline tests/testthat/test-reporter-junit.R]: Will do. @jspitzen do you think you could help out a bit?

Comment (jspitzen) [inline tests/testthat/test-reporter-junit.R]: Sure, I can help out. I don't have a ton of free time. I'll look at what has happened since I last checked this PR tonight and see what I can do.

Comment (lbartnik) [inline tests/testthat/test-reporter-junit.R]: I thought mostly of tests since I know little about right and wrong when it comes to junit output.

**Comment by lbartnik** [comment]
Hi @jspitzen ,@hadley,

I've added a regression test and fixed a minor issue with the reporter. I'm working on moving from `XML` to `xml2` but R sessions becomes very unstable when I use `xml2` to create the document. This is what happens when I try to test JUnit reporter in RStudio:

```
old-school` expectations: ...............
JunitReporter: ..
ListReporter: .......
MultiReporter: 
Exited with status -1073740940.
```

Lukasz

**Comment by lbartnik** [comment]
It seems `xml2` works now. Let me know what you think.

**Comment by lbartnik** [comment]
Strange... R CMD check works in my system.

```
_R_CHECK_CRAN_INCOMING_=FALSE /usr/bin/R CMD check testthat_1.0.2.9000.tar.gz --no-manual --as-cran --install-args=--build --no-multiarch
Running R from: /usr/lib/R 
* using log directory ‘/home/lukaszb/Workspace/testthat.Rcheck’
* using R version 3.3.1 (2016-06-21)
* using platform: x86_64-pc-linux-gnu (64-bit)
* using session charset: UTF-8
* using options ‘--no-manual --as-cran’
....
* checking tests ...
  Running ‘test-catch.R’ [11s/12s]
  Running ‘testthat.R’
 OK
* DONE

Status: 1 NOTE
See
  ‘/home/lukaszb/Workspace/testthat.Rcheck/00check.log’
for details.
```

**Comment by dirkschumacher** [comment]
@lbartnik It seems like at some point on Windows the package Rcpp is tried to be loaded:
```
** preparing package for lazy loading
Error in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]) : 
  there is no package called 'Rcpp'
ERROR: lazy loading failed for package 'testthat'
```

One quick fix would be to change `appveyor.yml`:
```yml
build_script:
  - travis-tool.sh install_deps
  - travis-tool.sh install_r Rcpp
```

However I guess it would make more sense to investigate where the Rcpp dependency comes from.

**Comment by lbartnik** [comment]
@dirkschumacher thanks for looking into that. Apparently xml2 imports Rcpp. I'm not sure how does this build work but shouldn't it download all dependency tree in the first place? I'm not sure if there is anything I can do, apart from specifying Rcpp explicitly in testthat - but that doesn't seem to be a good idea, even if it solves anything.

**Comment by dirkschumacher** [comment]
Maybe @hadley knows why `Rcpp` is not installed as a dependency of `xml2` on Windows/Appveyor?

**Comment by sebastian-c** [comment]
I've tested this fork and it works, only catch is for anyone who was using the other fork by @jspitzen is that the name of the reporter has changed from `JUnitReporter` to `JunitReporter`.

**Comment by lbartnik** [comment]
@dirkschumacher's advice has proven correct. AppVeyor doesn't choke on tests anymore :-)

**Comment by hadley** [comment]
Could you please merge/rebase to clean up the commit history?  Then I'll do a more thorough review.  

Also need to add to `test-reporter.R` for regression tests.  

I'm planning for a release early next year so it would be great if we can (finally) get this in.

**Comment by lbartnik** [comment]
I've merged but I'm not sure if that's enough to "clean up the commit history". Can you elaborate what exactly would be your goal? I could copy the current diff into a new branch and that way you'd have a single commit with no history there - but I thought you would rather keep track of all actual contributors.

I've also added a few lines to `test-reporter.R` but now I'm not sure if the current design is correct. By default this reporter writes to a file `./report.xml` and the standard output gets something much more simplified and human-readable. But maybe the XML should be displayed directly on stdout and the file should be an option that's not used by default?

File: R/reporter-junit.R [hunk 3]
```diff
@@ -0,0 +1,163 @@
+#' @include reporter.R
+NULL
+
+classnameOK <- function(text) {
+  gsub("[ \\.]", "_", text)
+}
+
+#' @importFrom xml2 xml_attr<- xml_add_child
+add_broken <- function (parent, type, message) {
```
Comment (hadley) [inline R/reporter-junit.R]: I'd prefer that you not import these functions and instead use fully qualified calls.

File: R/reporter-junit.R [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  }
+}
+
+
+#' Test reporter: summary of errors in jUnit XML format.
+#'
+#' This reporter includes detailed results about each test and summaries,
+#' written to a file (or stdout) in jUnit XML format. This can be read by
+#' the Jenkins Continuous Integration System to report on a dashboard etc.
+#' Requires the XML package.
```
Comment (hadley) [inline R/reporter-junit.R]: "xml2"

File: R/reporter-junit.R [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+#'
+#' This works for \code{\link{expect_that}} but not for the wrappers like
```
Comment (hadley) [inline R/reporter-junit.R]: Why not?

File: R/reporter-junit.R [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+#' The deparsed expect_that() call becomes the \code{<testcase>} name.
+#' On failure, the message goes into the \code{<failure>} node message
+#' argument (first line only) and into its text content (full message).
+#'
+#' Execution time and some other details are also recorded.
+#'
+#' References for the jUnit XML format:
+#' \url{http://llg.cubic.org/docs/junit/}
+#'
+#' Output drawn from the SummaryReporter is printed to the standard
```
Comment (hadley) [inline R/reporter-junit.R]: One reporter should do one thing, so I agree that you should drop this and simply output xml

File: R/reporter-junit.R [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    },
+
+    add_result = function(context, test, result) {
+      self$tests <- self$tests + 1
+
+      time <- self$elapsed_time()
+      self$suite_time <- self$suite_time + time
+
+      # XML node for test case
+      testcase <- xml_add_child(self$suite, "testcase")
```
Comment (hadley) [inline R/reporter-junit.R]: Can eliminate `add_attrs` in favour of:

```R
testcase <- xml_add_child(self$suite, 
  "testcase",
  time = time,
  classname = paste0(classnameOK(context), '.', classnameOK(name)))
)
```

(And please note my preferred indenting style)

File: R/reporter-junit.R [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+      name <- if (is.null(test) || !nchar(test)) "(unnamed)" else test
```
Comment (hadley) [inline R/reporter-junit.R]: Please use `name <- test %||% "(unnamed)"`

File: R/reporter-junit.R [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        message  <- paste(as.character(result), location)
+      }
+
+      # add child XML node if not success
+      if (expectation_error(result)) {
+        self$cat_tight(single_letter_summary(result))
+        add_broken(testcase, 'error', message)
+        self$errors <- self$errors + 1
+      }
+      else if (expectation_failure(result)) {
```
Comment (hadley) [inline R/reporter-junit.R]: Please more `else` to previous line

File: R/reporter-junit.R [hunk 10]
```diff
@@ -0,0 +1,163 @@
+#' @include reporter.R
+NULL
+
+classnameOK <- function(text) {
+  gsub("[ \\.]", "_", text)
+}
+
+#' @importFrom xml2 xml_attr<- xml_add_child
+add_broken <- function (parent, type, message) {
+  child <- xml_add_child(parent, type)
```
Comment (hadley) [inline R/reporter-junit.R]: You can assign the attributes during creation

File: appveyor.yml [hunk 11]
```diff
@@ -14,6 +14,7 @@ install:
 
 build_script:
   - travis-tool.sh install_deps
+  - travis-tool.sh install_r Rcpp
```
Comment (hadley) [inline appveyor.yml]: Why is this necessary?

File: R/reporter-junit.R (hunk 3 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Can I ask why?

File: R/reporter-junit.R (hunk 7 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: R/reporter-junit.R (hunk 4 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: R/reporter-junit.R (hunk 10 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: R/reporter-junit.R (hunk 8 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: appveyor.yml (hunk 11 above)
Comment (lbartnik) [inline appveyor.yml]: AppVeyor's build would fail on Windows without this - Rcpp wouldn't be installed. I don't know why but this fixed the problem.

File: R/reporter-junit.R (hunk 7 above)
Comment (lbartnik) [inline R/reporter-junit.R]: By the way, I used the default RStudio indentation setting.

File: R/reporter-junit.R (hunk 9 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: NAMESPACE [hunk 12]
```diff
@@ -131,6 +132,10 @@ importFrom(crayon,green)
 importFrom(crayon,red)
 importFrom(crayon,yellow)
 importFrom(magrittr,"%>%")
+importFrom(xml2,"xml_attr<-")
```
Comment (hadley) [inline NAMESPACE]: These can go away now

File: R/reporter-junit.R [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+      self$doc   <- xml2::xml_new_document()
+      self$root  <- xml2::xml_add_child(self$doc, 'testsuites')
+      self$reset_suite()
+    },
+
+    start_context = function(context) {
+      self$cat(context, ": ")
+
+      self$suite <- xml2::xml_add_child(self$root, "testsuite")
+      add_attrs(self$suite,
```
Comment (hadley) [inline R/reporter-junit.R]: Could now be inlined into call above, and `add_attrs()` deleted, right?

File: appveyor.yml (hunk 11 above)
Comment (hadley) [inline appveyor.yml]: @krlmlr any ideas?

File: R/reporter-junit.R [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+        message  <- paste(as.character(result), location)
+      }
+
+      # add child XML node if not success
+      if (expectation_error(result)) {
+        self$cat_tight(single_letter_summary(result))
+        add_broken(testcase, 'error', message)
+        self$errors <- self$errors + 1
+      } else
+      if (expectation_failure(result)) {
```
Comment (hadley) [inline R/reporter-junit.R]: `if` should be on same line as `else`

File: R/reporter-junit.R [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+      message <- if (is.null(result$call)) "(unexpected)" else format(result$call)[1]
+
+      if (!is.null(result$srcref)) {
+        location <- paste0('@', attr(result$srcref, 'srcfile')$filename, '#', result$srcref[1])
+        message  <- paste(as.character(result), location)
+      }
+
+      # add child XML node if not success
+      if (expectation_error(result)) {
+        self$cat_tight(single_letter_summary(result))
```
Comment (hadley) [inline R/reporter-junit.R]: I don't think you should be printing these out at all

File: R/reporter-junit.R [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        add_broken(testcase, 'error', message)
+        self$errors <- self$errors + 1
```
Comment (hadley) [inline R/reporter-junit.R]: And I don't think you're using `errors` anywhere, so you can remove.

File: R/reporter-junit.R [hunk 17]
```diff
@@ -0,0 +1,153 @@
+#' @include reporter.R
+NULL
+
+classnameOK <- function(text) {
+  gsub("[ \\.]", "_", text)
+}
+
+add_broken <- function (parent, type, message) {
```
Comment (hadley) [inline R/reporter-junit.R]: Don't need assignment any more, and maybe should be `add_failure_node()`?  Are you sure the `type` is both the node name and an attribute?

File: R/reporter-junit.R [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+      {
+        self$cat_tight(colourise(".", "success"))
+      }
+    },
+
+    end_reporter = function() {
+      self$cat("\n")
+
+      # this causes a segfault write_xml(xmlDoc, self$file, format = )
+      cat(toString(self$doc), file = self$file)
```
Comment (hadley) [inline R/reporter-junit.R]: This should just be printed to stdout like all the other reporters.

File: NAMESPACE (hunk 12 above)
Comment (lbartnik) [inline NAMESPACE]: Done.

File: R/reporter-junit.R (hunk 17 above)
Comment (lbartnik) [inline R/reporter-junit.R]: I followed the previous author here but type should be either the "exception class" (for error) or "type of the assertion" for failure). However, it doesn't seem we have an equivalent value here and XSD requires a type. Do you have a better idea what to put there?

File: R/reporter-junit.R (hunk 13 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: R/reporter-junit.R (hunk 15 above)
Comment (lbartnik) [inline R/reporter-junit.R]: That's deleted in my branch...

File: R/reporter-junit.R (hunk 16 above)
Comment (lbartnik) [inline R/reporter-junit.R]: I do use them, in `end_context`

File: R/reporter-junit.R (hunk 18 above)
Comment (lbartnik) [inline R/reporter-junit.R]: What if I'd like to output XML to a file? Do you expect people to `capture.output()` ?

**Comment by lbartnik** [comment]
OK, waiting for the next batch of comments... ;-)

File: R/helper-junitmock.R [hunk 19]
```diff
@@ -0,0 +1,18 @@
+createJunitReporterMock <- function (file) {
```
Comment (hadley) [inline R/helper-junitmock.R]: Could you add a brief comment along the lines of "Fix components of JunitReporter that otherwise vary from run-to-run"

File: R/reporter-junit.R [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+        self$failures <- self$failures + 1
+      } else if (expectation_skip(result)) {
+        xml2::xml_add_child(testcase, "skipped")
+        self$skipped <- self$skipped + 1
+      }
+    },
+
+    end_reporter = function() {
+      if (inherits(self$file, "connection")) {
+        file <- tempfile()
```
Comment (hadley) [inline R/reporter-junit.R]: Would you mind filing an issue for this on xml2 as well?  (It's fine to leave this workaround as is, but long-term xml2 should also supporting writing to connections)

File: R/reporter-junit.R [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+        xml2::write_xml(self$doc, file, format = TRUE)
+        writeLines(readLines(file), self$file)
+      } else if (is.character(self$file)) {
+        xml2::write_xml(self$doc, self$file, format = TRUE)
+      }
+      else {
```
Comment (hadley) [inline R/reporter-junit.R]: Should be on same line as previous.

File: tests/testthat/test-reporter-junit.R [hunk 22]
```diff
@@ -0,0 +1,27 @@
+context("JUnitReporter")
+
+
+test_that("Junit reporter regression", {
```
Comment (hadley) [inline tests/testthat/test-reporter-junit.R]: Do you think this is needed as well as the saved xml file? It seems a bit redundant to me.

File: R/helper-junitmock.R (hunk 19 above)
Comment (lbartnik) [inline R/helper-junitmock.R]: Done.

File: R/reporter-junit.R (hunk 20 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Submitted under: https://github.com/hadley/xml2/issues/157

File: tests/testthat/test-reporter-junit.R (hunk 22 above)
Comment (lbartnik) [inline tests/testthat/test-reporter-junit.R]: No, you're right.

File: R/reporter-junit.R (hunk 21 above)
Comment (lbartnik) [inline R/reporter-junit.R]: Done.

File: appveyor.yml (hunk 11 above)
Comment (jimhester) [inline appveyor.yml]: Rcpp is not being installed probably due to https://github.com/hadley/devtools/issues/1246 or possibly https://github.com/hadley/devtools/issues/1409. Use devel devtools in appveyor should fix it.

File: R/helper-junitmock.R [hunk 23]
```diff
@@ -0,0 +1,26 @@
+# Fix components of JunitReporter that otherwise vary from run-to-run
```
Comment (hadley) [inline R/helper-junitmock.R]: Oh I just noticed this in `R/` not `tests/testthat`. Can you please move there? (And then we're done 🎉)

**Comment by lbartnik** [comment]
It seems I don't need the `tempfile()` workaround but then XML won't be pretty-formatted. https://github.com/hadley/xml2/issues/157

File: R/helper-junitmock.R (hunk 23 above)
Comment (lbartnik) [inline R/helper-junitmock.R]: Oh, shoot... Sorry for that.

Comment (lbartnik) [inline R/helper-junitmock.R]: Done.

**Comment by hadley** [comment]
Sorry, I forgot to ask for a bullet in `NEWS.md` — can you please add?

**Comment by lbartnik** [comment]
Sure. Is that OK?

**Comment by hadley** [comment]
Perfect - thanks!
